woocommerce/woocommerce-android
# Add support for accepting payments - capturePayment not supported

Parent issue #3726

Merge instructions
1. Make sure #3780 is merged
2. Remove "Not ready for merge" label
3. Merge this PR

This PR adds support for collecting card payments using a card reader.

There are four necessary steps the app/user needs to make to successfully collect a payment:
1. CreatePaymentIntent
2. CollectPaymentMethod
3. ProcessPayment
4. Capture Payment

This PR introduces 4 classes with `Action` suffix. Each of these classes is responsible for one of the steps. The classes transform callbacks coming from StripeTerminalSDK into CallbackFlows to make the code a bit more readable. 
One of the most interesting "Action" is "CollectPaymentAction". This class needs to wait for user input "tapping/inserting card" -> the StripeTerminalSDK invokes "onRequestReaderInput` callback, the "Action" class emits "WaitingForInput" event but doesn't finish the flow just yet -> the user taps the card -> StripeTerminalSDK invokes "onSuccess" callback -> the "Action" class emits "Success" and "close()" method which terminates the flow aka "collect payment action".

CardPaymentManager is responsible for orchestration of the steps mentioned above -> in other words, it invokes one step after another unless the step fails. This class will will in upcoming PRs  store the last successful step and will support retry for each of the steps. This has not been implemented yet.

This PR also introduces "Collect payment" button on OrderDetail screen and displays dummy snackbar messages as the payment flow progresses - available only when CardReader feature flag is enabled. All the changes in the fragment and related classes shouldn't have any user facing impact in production builds - please review this part carefully.

The last change is removal of unused classes - all these classes were living in "temporary" package and were not being used after these changes.

Note: I plan to add unit tests for all these classes in upcoming PRs. If you'd prefer me to include them in this PR so you can use them as documentation/example usage, please let me know. 

To test:
Testing this flow requires a hw reader and Stripe testing card. If you don't have access to it, please feel free (encouraged) to still review this PR and let me know. We can ask for example folks from iOS to help us testing the changes.

1. Connect to a store which has an active and working WCPay (in dev/test mode)
2. Open app settings (tap on gear icon in the top right corner)
3. Tap on "Manage card readers" (Make sure bluetooth is on)
4. Leave the "simulated card reader" unchecked
5. Tap on "Connect card reader" 
6. Notice "Started" snackbar
7. Turn on the card reader
8. Notice a textview below the "Connect card reader" button changes to "Connecting/Connected" (it might take up to 20s)
9. Leave the app settings and go to Orders
10. Open detail of an order
11. Tap on "Collect payment" button
12. Notice "Payment flow started" and "Tap your card" messages
13. Tap your card on the reader - wait until it beeps and remove your card
14. Notice "Processing payment" message -> the last message is "Payment failed" because the "capturePayment" method is not implemented yet. It's coming in the follow up PRs.



Update release notes:
